### PR TITLE
Skip setting hash if hash was not set previously

### PR DESCRIPTION
**Pull request by thomaschaaf**
**Summary**

This should fix https://github.com/yarnpkg/yarn/issues/1834 where yarn had resolved the hosted-git-packages without a hash and downloaded them to the cache folder `<Packagename>-<Version>`. The tarball fetcher would then add the hash after the download completed resulting in the linker believing that the Package could be found in the cache folder `<Packagename>-<Version>-<Hash>`.

This https://github.com/yarnpkg/yarn/commit/ba5a20307992057c49663fce2569c01f59266701 commit seems to have added the logic behind tarball hashes.

**Test plan**

I added end to end tests and some jest tests
yarn add mulesoft/api-console.git#master would not work. Now works.

### CODE REVIEW

**Comment by bestander** [comment]
I would appreciate a unit test here.
It should be a matter of adding a test similar to this https://github.com/yarnpkg/yarn/blob/master/__tests__/commands/add.js#L554 that just checks that dependency gets installed correctly.

**Comment by thomaschaaf** [comment]
I will add unit tests now :) Then we can hopefully merge it very soon

**Comment by bestander** [comment]
Thanks, @thomaschaaf, I'll be ready to merge it at once.

File: src/package-fetcher.js [hunk 1]
```diff
@@ -88,7 +88,11 @@ export default class PackageFetcher {
         newPkg = res.package;
 
         // update with new remote
-        ref.remote.hash = res.hash;
+        // but only if there was a hash previously as the tarball fetcher does not provide a hash.
+        if(ref.remote.hash) {
```
Comment (wyze) [inline src/package-fetcher.js]: Just a minor nit to add a space here, `if(` -> `if (`.

**Comment by bestander** [comment]
.tgz files got corrupted

**Comment by thomaschaaf** [comment]
I don't know if this is the correct way to write jest tests as this makes a real http request to github and downloads the files but I did test the unit tests and when I reverted the change the unit tests failed.

File: end_to_end_tests/README.md [hunk 2]
```diff
@@ -3,3 +3,4 @@ This directory contains end-to-end tests for Yarn. The tests perform the followi
 1. Spin up a fresh Docker container
 2. Install Yarn through the package repository
 3. Run `yarn add react` in a test directory
+4. Run `yarn add 'substack/node-mkdirp#master'` in a test directory
```
Comment (bestander) [inline end_to_end_tests/README.md]: let's remove this now that it is covered with unit test.

File: end_to_end_tests/data/run-yarn-test.sh [hunk 3]
```diff
@@ -17,3 +17,4 @@ mkdir yarntest
 cd yarntest
 echo {} > package.json
 yarn add react || fail_with_log
+yarn add 'substack/node-mkdirp#master' || fail_with_log
```
Comment (bestander) [inline end_to_end_tests/data/run-yarn-test.sh]: and this

**Comment by wyze** [comment]
Should there be some assertions in the unit test to ensure the module installed in the `node_modules` folder?

**Comment by bestander** [comment]
Yes, ideally there should be assertations.
Although install/add commands run check afterwards automatically, it would fail without the fix in place.

**Comment by bestander** [comment]
Does the test fail without the fix, @thomaschaaf ?

**Comment by thomaschaaf** [comment]
@bestander yes the tests fail if it is not fixed

**Comment by bestander** [comment]
I'll do 0.17.3 in 30 min

**Comment by wyze** [comment]
Thanks for the iterations @thomaschaaf!

**Comment by emveeoh** [comment]
thanks for the quick fix everyone! :)

**Pull request merged**